src/librustc/ty/layout.rs

```diff
@@ -744,7 +744,17 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                         else { StructKind::AlwaysSized }
                     };
 
-                    let mut st = univariant_uninterned(&variants[v], &def.repr, kind)?;
+                    let mut st = match tcx.lang_items().va_list() {
+                        Some(did) if did == def.did => {
+                            let variants = tcx.va_list_types().iter()
+                                                              .map(|x| self.layout_of(x).unwrap())
+                                                              .collect::<Vec<_>>();
+                            univariant_uninterned(&variants[..], &def.repr, kind)?
```

Comment:
I don't think this is entirely correct. I think this will not be correct for the `CharPtr` and `VoidPtr` variants.